src/DOM/DOM.js

```diff
@@ -129,30 +127,17 @@ function setDOMProps (element, name, value, xmlns) {
 		case 'width':
 		case 'height':
 			if (element.type === 'img')
-				break
-		default:
-			if (!xmlns && name in getDOMNode(element))
-				return setDOMPropsValue(element, name, value, true)
+				return setDOMAttribute(element, name, value, '')
 	}
 
-	setDOMPropsValue(element, name, value, false)
-}
-
-/**
- * @param {Element} element
- * @param {string} name
- * @param {*} value
- * @param {boolean} signature
- */
-function setDOMPropsValue (element, name, value, signature) {
 	switch (typeof value) {
+		case 'object':
+			return setDOMProperty(element, name, value && assign({}, getDOMNode(element)[name], value))
```

Comment:
why `assign()` to previous value? (vs replacing with new value)

Reply:
The new* value is a delta of what has changed for example in

```js
render(h('h1', {custom: {first: 1}}, '0'), container)
render(h('h1', {custom: {first: 1, second: 2}}, '0'), container)
```

The delta is `{second: 2}`

Reply:
I see. (I wonder if there's a perf difference caused by reading the value out of the DOM)
How does this affect non-plain objects?

```js
h('custom', {stream: new Rx.Observable.of('x')})
```

Reply:
>I wonder if there's a perf difference caused by reading the value out of the DOM

Entirely dependent on how the web component author has implemented it.

>How does this affect non-plain objects?

Non-enumerable properties are not traversable so there maybe no update diff if your non-plain object is made entirely of non-enumerable properties, though this still doesn't answer the question of how this also affects the creation phase.

What do you think?

Reply:
if the object is the same (by ref) - diff does nothing
else the object is different (by ref) and should replace the previous ref

Otherwise passing a new instance of a non-plain object doesn't work

Reply:
There some objects that we may only want to specifically apply the delta(i.e style objects). If two objects are deepEqual the props diff will not generate a prop update with the delta. The props diff is agnostic to the target renderer and kind of property, i'd prefer to keep it this level of abstraction/separation.

i.e

```
// initial render
h('custom', {stream: new Rx.Observable.of('x')})

// this will not generate a prop update
h('custom', {stream: new Rx.Observable.of('x')})

// this will generate a prop update
h('custom', {stream: new Rx.Observable.of('y')})
```
Without changing the heuristics for what generates a prop diff. I think we can use the new* whole value instead of the delta when there is indeed a valid diff(when the two objects are not deepEqual) for this branch case.

Reply:
The problem is assuming deepEquals will differ

```js
class A {
  constructor (x) {
    Object.defineProperty(this, 'x', {value: x})
  }
  get () {
    return this.x
  }
}
a = new A('a')
b = new A('b')
a.get === b.get // true
```

Reply:
Yes this is a heuristic that applies to props as well if we can't iterate over the properties(enumerable's) we can not know what keys and values it has, thus it would always produce a noop diff in the update phase, in the above example `a.get`, `b.get` and `constructor` are also non enumerable.

Reply:
Can you address:
> Edit: I'm not sure I understand the rationale behind deep-diffing props at all. Shallow comparison seems optimal

Reply:
>Edit: I'm not sure I understand the rationale behind deep-diffing props at all. Shallow comparison seems optimal

Take for example the `style` DOM prop

```js
h('div', {style: {..., margin: 10, width: 10}})
h('div', {style: {..., margin: 10, width: 10}})
```
Nothing has changed the deep-diff generated would be empty, a shallow compare would instead generate an update for every value and this would happen every-time whether one prop or no values have changed – resulting potentially(depending on the renderer target) in expensive calls to update the native view. We could special case certain values, but we'd loose the platform/renderer agnostic aspect of the reconciler.

Another case is that this also allows other renderers to design their API's to make reasonable use of object values(shallow/deep) and have the guarantee that updates would be granular to only what has changed.

Reply:
ok, I can see an argument for deep-diffing plain objects
However, we shouldn't deep-diff non-plain objects. (let the custom reconciler handle efficient diffing of these)
How does that sound?

Reply:
I'm open to implementation details, though i'm worried that it would be near-impossible to effectively differentiate between plain objects and non-plain objects, for example non-plain objects could share the same prototype as a plain object or some non-plain objects could be intended to act like plain objects, how would you differentiate all of this and most notably how would you then extend this to work across realms. 

Given you can avoid pitfalls in creating non-plain objects(values that can change should not be non-enumerable) to make use of the deep-diff heuristic, would it still present reasonable benefit over the predictable nature of the current heuristic.

Reply:
lodash provides a relatively robust heuristic on plain-object-ness:
https://github.com/lodash/lodash/blob/master/isPlainObject.js

Reply:
Looks robust, the potential overhead for that doesn't look very good though.

That said i'm still worried about the case about intent, we could assume that the author did not intent to use  `new ImmutableObjectConstructor(...)` like a plain object, but what if that assumption is wrong? and the whether it is more complicated to explain and work with one heuristic vs the other, that is that non-plain prop object values are treated differently(shallow diff) to plain-objects(deep-diff), vs every object is treated identical(deep-diff'd) and enumerable properties are not visited.

As you can tell i'm leaning more towards the current heuristic of it's ok that non-enumarable properties do not affect the diff since the case of registering an intended changing value as non-enumerable looks like an anti-pattern that could be avoided with less overhead than the alternative.

Reply:
While possible, I'd venture that in most cases (99%+) developers do not instantiate objects as classes and expect them to behave in the same way.

The exception would be meta-programming with `Proxy()`, which is correctly identified by `_.isPlainObject()` (returns `true` for a proxied plain-object)

There is also a correctness argument to be made that users shouldn't (ever) need to know how diffing works, which would mean removing all deep-diff optimizations. (I think this goes a bit too far)